pkg/phlaredb/downsample/downsample.go

```diff
@@ -0,0 +1,252 @@
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/pyroscope/pkg/phlaredb/block"
+	schemav1 "github.com/grafana/pyroscope/pkg/phlaredb/schemas/v1"
+	"github.com/grafana/pyroscope/pkg/util/build"
+)
+
+type interval struct {
+	durationSeconds int64
+	shortName       string
+}
+
+var (
+	intervals = []interval{
+		{
+			durationSeconds: 5 * 60,
+			shortName:       "5m",
+		},
+		{
+			durationSeconds: 60 * 60,
+			shortName:       "1h",
+		},
+	}
+	inputSamplesHistogram = promauto.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "pyroscope_downsampler_input_profile_samples",
+			Help:    "The number of samples per profile before downsampling",
+			Buckets: prometheus.ExponentialBuckets(32, 2, 15),
+		})
+	outputSamplesHistogram = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "pyroscope_downsampler_output_profile_samples",
+			Help:    "The number of samples per profile after downsampling",
+			Buckets: prometheus.ExponentialBuckets(32, 2, 15),
+		}, []string{"interval"})
+)
+
+type profilesWriter struct {
+	*parquet.GenericWriter[*schemav1.Profile]
+	file *os.File
+
+	buf []parquet.Row
+}
+
+func (p *profilesWriter) WriteRow(r parquet.Row) error {
+	p.buf[0] = r
+	_, err := p.GenericWriter.WriteRows(p.buf)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func newProfilesWriter(path string, i interval, aggregation string) (*profilesWriter, error) {
+	profilePath := filepath.Join(path, fmt.Sprintf("profiles_%s_%s", i.shortName, aggregation)+block.ParquetSuffix)
+	profileFile, err := os.OpenFile(profilePath, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o644)
+	if err != nil {
+		return nil, err
+	}
+	return &profilesWriter{
+		GenericWriter: newParquetProfileWriter(profileFile, parquet.MaxRowsPerRowGroup(100_000)),
+		file:          profileFile,
+		buf:           make([]parquet.Row, 1),
+	}, nil
+}
+
+func newParquetProfileWriter(writer io.Writer, options ...parquet.WriterOption) *parquet.GenericWriter[*schemav1.Profile] {
+	options = append(options, parquet.PageBufferSize(3*1024*1024))
+	options = append(options, parquet.CreatedBy("github.com/grafana/pyroscope/", build.Version, build.Revision))
+	options = append(options, schemav1.DownsampledProfilesSchema)
+	return parquet.NewGenericWriter[*schemav1.Profile](
+		writer, options...,
+	)
+}
+
+type aggregationType struct {
+	fn   func(a, b int64) int64
+	name string
+}
+
+var (
+	SumAggregation = aggregationType{
+		fn: func(a, b int64) int64 {
+			return a + b
+		},
+		name: "sum",
+	}
+)
+
+type Downsampler struct {
+	aggregation    aggregationType
+	path           string
+	profileWriters []*profilesWriter
+	states         []*state
+}
+
+type state struct {
+	currentRow          parquet.Row
+	currentTime         int64
+	currentFp           model.Fingerprint
+	totalValue          int64
+	stackTraceIds       []uint64
+	values              []int64
+	stackTraceIdToIndex map[uint64]int //
```

Comment:
The order of samples in compacted blocks [is not deterministic](https://github.com/grafana/pyroscope/blob/main/pkg/phlaredb/compact.go#L725), and this may cause encoding efficiency degradation, indeed. This is definitely something we should play with